core/src/main/java/software/amazon/awssdk/auth/DefaultCredentialsProvider.java

```diff
@@ -64,13 +66,18 @@ private DefaultCredentialsProvider(Builder builder) {
      */
     private static AwsCredentialsProviderChain createChain(Builder builder) {
         AwsCredentialsProvider[] credentialsProviders = new AwsCredentialsProvider[] {
-                new SystemPropertyCredentialsProvider(),
-                new EnvironmentVariableCredentialsProvider(),
-                new ProfileCredentialsProvider(),
-                ElasticContainerCredentialsProvider.builder()
-                                                   .asyncCredentialUpdateEnabled(builder.asyncCredentialUpdateEnabled)
-                                                   .build(),
-                InstanceProfileCredentialsProvider.builder()
+            new SystemPropertyCredentialsProvider(),
+            new EnvironmentVariableCredentialsProvider(),
+            new ProfileCredentialsProvider(),
+            ContainerCredentialsProvider.builder()
+                                        .credentialsEndpointProvider(new EcsCredentialsEndpointProvider())
+                                        .asyncCredentialUpdateEnabled(builder.asyncCredentialUpdateEnabled)
+                                        .build(),
+            ContainerCredentialsProvider.builder()
+                                        .credentialsEndpointProvider(new FullUriGenericCredentialsEndpointProvider())
```

Comment:
Random idea: instead of having these 2 different credentials provider instances with their own endpoint providers (essentially 3 credential providers, since the instance profile credentials provider is similar) maybe we could have a "endpoint provider chain" with the three different endpoints, and a single container/instance profile credentials provider? I added some complexity earlier to creating the InstanceProfileCredentialsProvider / ContainerCredentialsProviders, and we're adding a bit more here. I'm worried it might be too hard for the customer to create one of these themselves now.

More simply, we could just have a `ContainerType` enum instead of the endpoint provider with ECS/EC2/DEFAULT that we internally map to the so the customer doesn't need to find the right endpoint provider implementation.

Reply:
+1 to some simplification here. Seems like these are all just hitting different endpoints to retrieve credentials.

Reply:
Will take a look